compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/LocalDeclarationsLowering.kt

```diff
@@ -744,15 +744,17 @@ class LocalDeclarationsLowering(
 
                     assert(declaration.visibility != Visibilities.LOCAL)
 
-                    if ((declaration.parent as IrClass).isInner) return
+                    val parent = declaration.parent as IrClass
+                    if (parent.visibility != Visibilities.LOCAL || parent.isInner) return
 
                     localClassConstructors[declaration] = LocalClassConstructorContext(declaration)
                 }
 
                 override fun visitClass(declaration: IrClass) {
                     super.visitClass(declaration)
 
-                    if (declaration.isInner) return
+                    // Inner classes inside local classes have local visibility for some reason.
+                    if (declaration.visibility != Visibilities.LOCAL || declaration.isInner) return
```

Comment:
Actually LocalDeclarationLower was rewriten a bit and would be merged today: see rr/mb/ldl branch, this commit would be redundant

Reply:
I'll wait until it is merged then, one of the tests unmuted here is dependent on this change plus another one.

Reply:
@pyos  Just merged